### PR TITLE
Add tokenization and prompting API to GPT models

### LINKED ISSUE #646: Add tokenization support to Disco LLMs
Previous and current works on LLM integration to DISCO relies on pre-tokenized datasets and doesn't account for token decoding after inference.

Full tokenizer support would allow:
- [x] To input natural text which is tokenized in DISCO
- [x] To perform Token decoding after model inference
- [x] To be able to use a pre-trained LLM in Disco first requires converting the weights to a format compatible with TF.js or other JS libraries. But additionally, we will also need to use the model's "pre-trained" tokenizer, and will also need to convert the tokenizers to JavaScript.
- [ ] Train a tokenizer with a particular algorithm (e.g. SentencePiece BPE) on a custom dataset, use it when training a model, and save it along with the model
- [ ] To save and store tokenizer alongside the model with which it was used

### PR DESCRIPTION
**Pull request by JulienVig**
Fixes #646 

* Adds tokenization support, allowing to choose any pre-trained tokenizer from Transformers.js
* Implement text generation with a trained model
* Extend text processing with shuffling and left padding
* Create an example of language model training in `docs/examples/wikitext.ts`
* I encountered a lot of issues and limitations which I listed in #654

### CODE REVIEW

**Comment by JulienVig** [comment]
@tharvik @peacefulotter I'd be happy to hear your take on how to integrate tokenization and prompting into the Disco architecture.
The initial commit is a POC of a pipeline from text training data (so not already preprocessed) to textual output (cf `docs/example/wikitext.ts`)
I wanted to make the tokenizer as a `trainingInformation` field in the infamous TaskProvider but it seems that complex objects (like functions) are removed from the task object when communicated between the server and the client...
My next step is to migrate away from `gpt-tokenizer` to use [`Transformers.js`](https://huggingface.co/docs/transformers.js/api/tokenizers) such that we can specify which pre-trained tokenizer we want to use as a string in the trainingInformation

**Comment by peacefulotter** [comment]
Congrats on the LLM POC milestone!

Firstly, I am wondering if having a custom tokenizer is even really relevant in the first place. Is there really a use case for it in Disco?

If so, instead of passing the tokenizer object, you could pass a string representing the tokenizer to use. Then an object could map tokenizer ids / names to the corresponding instance.

**Comment by peacefulotter** [comment]
Secondly, since it appears tokenization needs to be done on the fly (before or during training) and not as a completely separate step. It could be done as a preprocessing step like Hugo planned to do on his branch

**Comment by JulienVig** [comment]
@peacefulotter yes I want to do both points you mentioned! The second one is already implemented, the tokenization is part of the preprocessing.
And for the first I want to migrate to Transformers.js for that reason, it allows loading different pre-trained tokenizer from a string (rather than an import like gpt-tokenizer)

**Comment by tharvik** [comment]
> I'd be happy to hear your take on how to integrate tokenization and prompting into the Disco architecture.

I don't like `Task`, it's the goto place for every context variable in Disco. here are some questions before adding it to task
  * is it applicable to every model? ie, can images be tokenized?
  * would someone want to change the tokenizer of wikitext?

after a bit more thinking, I don't think it really matters for now, as long as #639 is not done

> I wanted to make the tokenizer as a `trainingInformation` field in the infamous TaskProvider but it seems that complex objects (like functions) are removed from the task object when communicated between the server and the client...

indeed, functions can't be represented by msgpack/json so it gets removed. it'll be nice to have a serialization process to ensure that we get function back from the network (a string such as `gpt-tokenizer/davinci` in the serialized object that get mapped to a known function)

**Comment by JulienVig** [comment]
Transformer.js requires Node >= 18, I'm waiting on #653 to be merged

File: web-client/.gitignore [hunk 1]
```diff
@@ -1,3 +1,4 @@
 # local env files
 .env.local
 .env.*.local
+node_modules
```
Comment (JulienVig) [inline web-client/.gitignore:4]: A node_modules folder kept getting created, I think when running the web-client's tests. I don't know if it's me that messed up with something or the cypress upgrade

File: discojs/discojs-core/package.json [hunk 2]
```diff
@@ -22,6 +22,7 @@
   "dependencies": {
     "@tensorflow/tfjs": "4",
     "@types/msgpack-lite": "0.1",
+    "@xenova/transformers": "^2.16.1",
```
Comment (tharvik) [inline discojs/discojs-core/package.json]: `npm i --save` likes to add the full version, but I do prefer to use the package-lock.json to keep track of the actual version we're supporting and keep package.json will only the major version

```suggestion
    "@xenova/transformers": "2",
```

I just noticed that "gpt-tokenizer" is also a dep, but was never used. can you also remove it?

File: discojs/discojs-node/src/data/text_loader.ts [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
-  async loadDatasetFrom (source: string): Promise<data.Dataset> {
-    // TODO sure, good idea to load the whole dataset in memory #irony
-    const content = await fs.readFile(source)
-    const file = new tfData.FileDataSource(content)
-
-    return new tfData.TextLineDataset(file)
+  loadDatasetFrom (source: string): Promise<data.Dataset> {
+    // TODO: reads all the file at once, 
+    // inputting the file path to FileDataSource isn't supported anymore
+    const inputFile = fs.readFileSync(source)
```
Comment (tharvik) [inline discojs/discojs-node/src/data/text_loader.ts]: better to use async `readFile` from 'node:fs/promises', as it allow for the other async methods to run while the file is loading.

```suggestion
    const inputFile = await fs.readFile(source)
```

File: discojs/discojs-core/src/task/training_information.ts [hunk 4]
```diff
@@ -50,6 +50,14 @@ export interface TrainingInformation {
   // aggregator:  aggregator to be used by the server for federated learning, or by the peers for decentralized learning
   // default is 'average', other options include for instance 'bandit'
   aggregator?: AggregatorChoice
+  // tokenizerName (string). For example: 'Xenova/gpt2'. The name should match a Transformers.js tokenizer available on HuggingFace's hub. 
+  tokenizerName?: string
+  // tokenizer (object). The actual tokenizer. It is initialized according to the tokenizerName the first time it is needed for the subsequent tokenizations
+  // This field is not expected to be filled when initializing a field
+  tokenizer?: object | null
```
Comment (tharvik) [inline discojs/discojs-core/src/task/training_information.ts]: I feel like both are linked and multually exclusive, did you try having it a single field?
also the only type of tokenizer we're currently supporting is a `PreTrainedTokenizer` (`object` is way too generic).
and I do recommend the use of `undefined` where possible instead of `null` (both are bottom types, why JavaScript..), as it's more "in the language" (partially init object, default args, … are undef)
smth like sounds nice: `tokenizer?: string | PreTrainedTokenizer`

File: discojs/discojs-core/src/dataset/data/preprocessing/text_preprocessing.ts [hunk 5]
```diff
@@ -1,57 +1,84 @@
...
+      tokenizer = await AutoTokenizer.from_pretrained(tokenizerName)
+      task.trainingInformation.tokenizer = tokenizer
+    }
+    const maxLength = task.trainingInformation.maxSequenceLength ?? tokenizer.model_max_length as number
+    // Should never happen because tokenization truncates inputs
+    if (xs.size > maxLength) {
+      xs = xs.slice([0], [maxLength])
+    } else if (xs.size < maxLength) {
+      const paddingToken = tokenizer.pad_token_id
+      xs = xs.pad([[Math.max(0, maxLength - xs.size), 0]], paddingToken)
```
Comment (tharvik) [inline discojs/discojs-core/src/dataset/data/preprocessing/text_preprocessing.ts:41]: ~~note there is [support for padding in transformer.js' tokenizers](https://github.com/xenova/transformers.js/blob/main/src/tokenizers.js#L2498). from what I saw in code, it's not allowing to left pad, only to right pad (still in a TODO state). consider adding a line mention to know where to look for updates.~~

and I read a bit further now, you already thought of that :)

File: discojs/discojs-core/src/models/gpt/evaluate.ts [hunk 6]
```diff
@@ -1,50 +1,58 @@
...
 
+interface DataPoint extends tf.TensorContainerObject {
+  xs: tf.Tensor2D,
+  ys: tf.Tensor3D,
+}
+
 export default async function evaluate (
   model: tf.LayersModel,
-  dataset: tf.data.Dataset<{ xs: tf.Tensor, ys: tf.Tensor }>
+  dataset: tf.data.Dataset<{ xs: tf.Tensor, ys: tf.Tensor }>,
```
Comment (tharvik) [inline discojs/discojs-core/src/models/gpt/evaluate.ts]: no need to cast anymore in here, making the function safer

```suggestion
  dataset: tf.data.Dataset<DataPoint>,
```

File: discojs/discojs-core/src/models/gpt/evaluate.ts [hunk 7]
```diff
@@ -1,50 +1,58 @@
...
   const acc: [number, number] = [0, 0]
 
-  await dataset.map(({ xs, ys }) => {
+  let iteration = 1
+  const iterator = await dataset.iterator()
+  let continueIterating = true
+  while (continueIterating) {
+    const next = await iterator.next()
+
+    const { xs, ys } = next.value as DataPoint
```
Comment (tharvik) [inline discojs/discojs-core/src/models/gpt/evaluate.ts]: the iterator is not infinite, you will [read an invalid `next.value`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterator_protocol) at the end of it (when `done` is true).
I don't really see the need of moving to a while loop instead of the stream/function interface: the latter takes care of the iteration for you, stopping when needed (use `dataset.take(maxEvalBatches)`).

File: discojs/discojs-core/src/models/gpt/evaluate.ts [hunk 8]
```diff
@@ -1,50 +1,58 @@
...
+    const lossValue = await lossTensor.array()
+    if (typeof lossValue !== 'number') {
       throw new Error('got multiple loss')
     }
 
     const accTensor = tf.metrics.categoricalAccuracy(ys, logits)
     const accSize = accTensor.shape.reduce((l, r) => l * r, 1)
     const accSum = accTensor.sum()
-    const accSummed = await accSum.array()
+    const accSummed = accSum.arraySync()
```
Comment (tharvik) [inline discojs/discojs-core/src/models/gpt/evaluate.ts]: but why? async is highly concurrent when doing IO (which might happen depending on the running platform)

File: docs/examples/wikitext.ts
Comment (tharvik) [inline docs/examples/wikitext.ts:1]: can you add a `script` in package.json for it? and run it in the `test` script

File: docs/examples/wikitext.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  const url = new URL('http://localhost:8080')
+  const tasks = await fetchTasks(url)
+  const task = tasks.get('wikitext-103')
+  if (task === undefined) { throw new Error('task not found') }
+  const dataset = await loadWikitextData(task)
+  const aggregator = new aggregators.MeanAggregator()
+  const client = new clients.federated.FederatedClient(url, task, aggregator)
+  const trainingInformant = new informant.FederatedInformant(task, 10)
+  const disco = new Disco(task, { scheme: 'federated', client, aggregator, informant: trainingInformant })
+  await disco.fit(dataset)
```
Comment (tharvik) [inline docs/examples/wikitext.ts]: it's a huge block of code which is not very understable to newcomers: can you add a few comments, new lines to make it least daunting plz?

File: web-client/.gitignore (hunk 1 above)
Comment (tharvik) [reply web-client/.gitignore:4]: yeah, I got it too, it's for a cache of some build files (as if we don't have temporary dir dedicated to that), it's not triggering anymore with the bump of vue, let's add it for now

**Review by tharvik — APPROVED** [review]
great work! a bit of nitpicking here and there but nothing really vital

File: discojs/discojs-core/src/models/gpt/evaluate.ts (hunk 7 above)
Comment (JulienVig) [reply discojs/discojs-core/src/models/gpt/evaluate.ts]: Yeah so I took the sad decision to use a while loop instead of the functional way because there was a memory leak when streaming over the dataset. I am still not exactly sure of where it happened but I suspect the iterator values were not correctly disposed. The while loop allows me to ensure it is done.
The check for `next.done` is done at the end of the loop, but it assumes at least one datapoint in the dataset, I'm adding an initial check on next.done before the while

File: discojs/discojs-core/src/models/gpt/evaluate.ts (hunk 8 above)
Comment (JulienVig) [reply discojs/discojs-core/src/models/gpt/evaluate.ts]: Good catch, it was not intentional 👍

File: discojs/discojs-core/src/task/training_information.ts (hunk 4 above)
Comment (JulienVig) [reply discojs/discojs-core/src/task/training_information.ts]: Yes good idea I'll try that now! The null was because of a linting error I wasn't sure how to solve but it should not be needed anymore

File: discojs/discojs-core/src/models/gpt/evaluate.ts (hunk 7 above)
Comment (JulienVig) [reply discojs/discojs-core/src/models/gpt/evaluate.ts]: Actually I thought you were talking about the training loop which I also converted from functional stream to an iterator loop and where the memory leak was occurring. 
There is still a growing number of allocated tensors (#654), especially in the evaluation loop, that's why I also tried changing the evaluate loop but it doesn't solve the issue. I'm reverting the evaluation loop to be functional.

File: discojs/discojs-core/src/models/tokenizer.ts
Comment (JulienVig) [inline discojs/discojs-core/src/models/tokenizer.ts:1]: @tharvik I've created a new `models/tokenizer.ts` file because code to load/initialize the tokenizer was getting duplicated. Let me know if see anything to improve!

Comment (JulienVig) [inline discojs/discojs-core/src/models/tokenizer.ts:1]: @tharvik I've created a new `models/tokenizer.ts` file because code to load/initialize the tokenizer was getting duplicated. Let me know if you see anything to improve!

File: discojs/discojs-core/src/models/gpt/evaluate.ts [hunk 10]
```diff
@@ -43,8 +49,46 @@ export default async function evaluate (
...
+
+  //   datasetSize += 1
+  //   totalLoss += lossValue
+  //   acc[0] += accSummed
+  //   acc[1] += accSize
+
+  //   tf.dispose([xs, ys, logits, accTensor, accSum, lossTensor, next.value])
+  //   iteration++
+  //   continueIterating = next.done !== true && iteration <= maxEvalBatches
+  // }
```
Comment (tharvik) [inline discojs/discojs-core/src/models/gpt/evaluate.ts]: let's not keep commented code around, it'll get out of sync

File: discojs/discojs-core/src/dataset/data/preprocessing/text_preprocessing.ts [hunk 11]
```diff
@@ -1,57 +1,74 @@
...
-const gpt3Tokenizer = null as unknown as { encode: (_: string) => { bpe: number[]; text: string[] } }
+/**
+ * We are currently only implementing left padding for text generation 
+ * https://huggingface.co/docs/transformers/en/llm_tutorial#wrong-padding-side
+ * The function can easily be extended to support right padding once the need arise
+ */
+const leftPadding: PreprocessingFunction = {
+  type: TextPreprocessing.LeftPadding,
+  apply: async (x: Promise<tf.TensorContainer>, task: Task): Promise<tf.TensorContainer> => {
+    let { xs } = await x as TokenizedEntry
```
Comment (tharvik) [inline discojs/discojs-core/src/dataset/data/preprocessing/text_preprocessing.ts:28]: just to be sure that people use it correctly, can you check that the `tf.TensorContainer` is really containing a field `xs`, of type `tf.Tensor` with `rankType === Rank.R1`?

File: discojs/discojs-core/src/dataset/data/preprocessing/text_preprocessing.ts [hunk 12]
```diff
@@ -1,57 +1,74 @@
...
+ * Tokenize and truncates input strings
+ */
 const tokenize: PreprocessingFunction = {
   type: TextPreprocessing.Tokenize,
-  apply: (x: tf.TensorContainer) => {
-    const { xs, ys } = x as TextEntry
-
-    const tokenized = gpt3Tokenizer.encode(xs[0]).bpe
+  apply: async (x: Promise<tf.TensorContainer>, task: Task): Promise<tf.TensorContainer> => {
+    const xs = await x as string // tf.TextLineDataset yields strings
```
Comment (tharvik) [inline discojs/discojs-core/src/dataset/data/preprocessing/text_preprocessing.ts:63]: same idea of checking that we indeed get `string` as input

**Review by tharvik — APPROVED** [review]
> I've created a new models/tokenizer.ts file

LGTM!

a few more comments (I'm a never ending stream of critics, please stop me)

File: discojs/discojs-core/src/models/gpt/evaluate.ts (hunk 10 above)
Comment (JulienVig) [reply discojs/discojs-core/src/models/gpt/evaluate.ts]: Good catch thanks! that was completely unintentional
